### ISSUE TITLE
Provisioning is performed immediately without waiting for the scripts to complete

### ISSUE THREAD

**Issue by salvax86** [test-kitchen/kitchen-ec2#389]
We run "kitchen converge" with the ec2 driver and ansible_push provisioning. To initialize the host for provisioning, we run a init script next to the cloud_init through user_data. 
Unfortunately, the provisioning does not wait for the user_data init to complete, resulting in an error. Is it possible to do something so that the provisioning step does not run into error?

.kitchen.yml
```
driver:
  name: ec2
  region: us-east-2
  associate_public_ip: true
  user_data: init.sh

platforms:
  - name: test-kitchen
    driver:
       image_id: ami-4f80b52a

transport:
  ssh_key: ./keys/ansible_kitchen.pem
  username: ubuntu
  connection_timeout: 10
  connection_retries: 5
  max_ssh_sessions: 6

provisioner:
  name: ansible_push
  ansible_config: ./test/ansible.cfg
  chef_bootstrap_url: nil
  idempotency_test: true
  diff: true

suites:
  - name: default
    provisioner:
      playbook: "./test/integration/default/default.yml"
    verifier:
      name: inspec
      patterns:
        - "./test/integration/default/rancher_spec.rb"
```

init.sh

```
#!/bin/bash

i="0"
while [ $i -lt 15 ]
do
if [ $(fuser /var/lib/dpkg/lock) ]; then
  i="0"
fi
sleep 1
i=$[$i+1]
done

sudo apt-get update;

sudo apt-get install curl python python-pip python-requests git libffi-dev libssl-dev -y;

pip install docker-py;
```

**Comment by cheeseplus** [comment]
This is sort of a general problem with user_data itself but the good news is that you're not the first to encounter it. 

I'd recommend reading [this](https://github.com/boto/boto/issues/2575#issuecomment-291450959) issue comment which hits the nail on the head with:

```
User data scripts are pretty much always custom things so you have to make some agreement between your user script and whatever is waiting for it. E.g. create a file when finished.
```
They then provide a helpful [link](https://stackoverflow.com/questions/11245356/how-to-check-user-data-status-while-launching-the-instance-in-aws/16578488#16578488) and [this ](https://serverfault.com/a/855898/66489)also seems promising.

In short, there isn't a ton kitchen-ec2 can do, especially when custom user_data is being passed. If there is something consistent and reliable across distros then I can see us perhaps adding a check but in general this is best left to the user.

**Comment by salvax86** [comment]
Thank you for the quick answer. 
Yes I read already these links before I ask here about it. At this moment my challenge is to use default (bare) amis for different plattform and before the ansible provisioner start, essential softwares (like ansible itself or python) should be already install. Now without install python or ansible I can not run ansible and I can not check if the script is complete or not. My workaround (at this moment) is to use an ami having all essential packages preinstalled.

**Comment by cheeseplus** [comment]
Indeed, the best way to do so is to have the provisioner (chef/ansible/etc) check itself because there really isn't a _good_ way kitchen can determine this as by that point all the code is _on_ the instance itself and kitchen is just getting the output. Pre-baking the AMIs like you are now is a great way to do solve this problem without worrying about user_data which I've (ab)used often.

I'm definitely thinking about ways in which we could try to prevent this but right now it seems like lots of sharp edge cases but I need to do some research to see if AWS gives us something semi-predictable to at least check.

**Comment by cheeseplus** [comment]
I'm closing this as it's a general problem with user_data/cloud-init like things and there isn't much we can do without lots of edge case hackery at this time. The workaround are solid and recommended for the time being.

**Comment by Zordrak** [comment]
You could add a custom wait. It would help as a workaround
